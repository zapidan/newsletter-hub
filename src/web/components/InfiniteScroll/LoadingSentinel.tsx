import React from 'react';
import { Loader2 } from 'lucide-react';

export interface LoadingSentinelProps {
  isLoading?: boolean;
  hasReachedEnd?: boolean;
  totalCount?: number;
  loadedCount?: number;
  error?: Error | null;
  onRetry?: () => void;
  className?: string;
}

/**
 * Loading sentinel component for infinite scroll
 * Displays loading state, end state, or error state at the bottom of the list
 */
export const LoadingSentinel: React.FC<LoadingSentinelProps> = ({
  isLoading = false,
  hasReachedEnd = false,
  totalCount,
  loadedCount,
  error,
  onRetry,
  className = '',
}) => {
  // Error state
  if (error) {
    return (
      <div className={`flex flex-col items-center justify-center py-8 text-center ${className}`}>
        <div className="text-red-500 mb-4">
          <svg
            className="w-12 h-12 mx-auto mb-2"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
            />
          </svg>
          <p className="text-sm font-medium">Failed to load more newsletters</p>
          <p className="text-xs text-gray-500 mt-1">
            {error.message || 'Something went wrong'}
          </p>
        </div>
        {onRetry && (
          <button
            onClick={onRetry}
            className="px-4 py-2 text-sm bg-red-50 hover:bg-red-100 text-red-700 rounded-lg transition-colors"
          >
            Try Again
          </button>
        )}
      </div>
    );
  }

  // End state
  if (hasReachedEnd) {
    return (
      <div className={`flex flex-col items-center justify-center py-8 text-center ${className}`}>
        <div className="text-gray-400 mb-2">
          <svg
            className="w-8 h-8 mx-auto"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M5 13l4 4L19 7"
            />
          </svg>
        </div>
        <p className="text-sm text-gray-500 font-medium">
          {totalCount !== undefined && loadedCount !== undefined
            ? `All ${totalCount} newsletters loaded`
            : 'No more newsletters to load'}
        </p>
        {totalCount !== undefined && loadedCount !== undefined && totalCount > 0 && (
          <p className="text-xs text-gray-400 mt-1">
            Showing {Math.min(loadedCount, totalCount)} of {totalCount} newsletters
          </p>
        )}
      </div>
    );
  }

  // Loading state
  if (isLoading) {
    return (
      <div className={`flex flex-col items-center justify-center py-8 ${className}`}>
        <Loader2 className="w-6 h-6 animate-spin text-blue-500 mb-3" />
        <p className="text-sm text-gray-500 font-medium">Loading more newsletters...</p>
        {totalCount !== undefined && loadedCount !== undefined && (
          <p className="text-xs text-gray-400 mt-1">
            Loaded {loadedCount} of {totalCount}
          </p>
        )}
      </div>
    );
  }

  // Default invisible sentinel for intersection observer
  return (
    <div
      className={`h-4 w-full ${className}`}
      aria-hidden="true"
    />
  );
};

export default LoadingSentinel;
