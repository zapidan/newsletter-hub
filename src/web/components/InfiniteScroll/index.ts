export { LoadingSentinel } from './LoadingSentinel';
export { InfiniteNewsletterList } from './InfiniteNewsletterList';
export type { LoadingSentinelProps } from './LoadingSentinel';
export type { InfiniteNewsletterListProps } from './InfiniteNewsletterList';

// Re-export infinite scroll hooks for convenience
export { useInfiniteScroll } from '../../../common/hooks/infiniteScroll/useInfiniteScroll';
export { useInfiniteNewsletters } from '../../../common/hooks/infiniteScroll/useInfiniteNewsletters';
export type {
  InfiniteScrollOptions,
  InfiniteScrollReturn
} from '../../../common/hooks/infiniteScroll/useInfiniteScroll';
export type {
  UseInfiniteNewslettersOptions,
  UseInfiniteNewslettersReturn
} from '../../../common/hooks/infiniteScroll/useInfiniteNewsletters';
