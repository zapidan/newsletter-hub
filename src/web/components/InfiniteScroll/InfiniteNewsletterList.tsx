import React, { useMemo } from "react";
import { NewsletterWithRelations, Tag } from "../../../common/types";
import { useInfiniteScroll } from "../../../common/hooks/infiniteScroll/useInfiniteScroll";
import NewsletterRow from "../NewsletterRow";
import { LoadingSentinel } from "./LoadingSentinel";

export interface InfiniteNewsletterListProps {
  newsletters: NewsletterWithRelations[];
  isLoading: boolean;
  isLoadingMore: boolean;
  hasNextPage: boolean;
  totalCount: number;
  error: Error | null;
  onLoadMore: () => void;
  onRetry?: () => void;

  // Newsletter row props
  selectedIds?: Set<string>;
  isSelecting?: boolean;
  readingQueue?: Array<{ newsletter_id: string }>;
  visibleTags?: Set<string>;

  // Newsletter row actions
  onRowClick?: (newsletter: NewsletterWithRelations) => void;
  onToggleSelect?: (id: string) => void;
  onToggleLike: (newsletter: NewsletterWithRelations) => Promise<void>;
  onToggleArchive?: (id: string) => void;
  onToggleRead?: (id: string) => void;
  onTrash?: (id: string) => void;
  onToggleQueue?: (newsletterId: string) => void;
  onUpdateTags?: (newsletterId: string, tagIds: string[]) => void;
  onToggleTagVisibility?: (id: string, e: React.MouseEvent) => void;
  onTagClick?: (tag: Tag) => void;
  onRemoveFromQueue?: (e: React.MouseEvent, newsletterId: string) => void;
  onMouseEnter?: (newsletter: NewsletterWithRelations) => void;

  // Loading states
  isDeletingNewsletter?: (id: string) => boolean;
  isUpdatingTags?: (id: string) => boolean;
  loadingStates?: Record<string, string>;

  // Error states
  errorTogglingLike?: Error | null;
  tagUpdateError?: string | null;
  onDismissTagError?: () => void;

  // Display options
  showTags?: boolean;
  showCheckbox?: boolean;

  // Infinite scroll options
  threshold?: number;
  rootMargin?: string;
  className?: string;
}

/**
 * Infinite scroll container for newsletters
 * Combines business logic from useInfiniteScroll with presentation logic
 * Renders a list of newsletters with automatic loading on scroll
 */
export const InfiniteNewsletterList: React.FC<InfiniteNewsletterListProps> = ({
  newsletters,
  isLoading,
  isLoadingMore,
  hasNextPage,
  totalCount,
  error,
  onLoadMore,
  onRetry,

  // Newsletter row props
  selectedIds = new Set(),
  isSelecting = false,
  readingQueue = [],
  visibleTags = new Set(),

  // Newsletter row actions
  onRowClick,
  onToggleSelect,
  onToggleLike,
  onToggleArchive,
  onToggleRead,
  onTrash,
  onToggleQueue,
  onUpdateTags,
  onToggleTagVisibility,
  onTagClick,
  onRemoveFromQueue,
  onMouseEnter,

  // Loading states
  isDeletingNewsletter = () => false,
  isUpdatingTags = () => false,
  loadingStates = {},

  // Error states
  errorTogglingLike,
  tagUpdateError,
  onDismissTagError,

  // Display options
  showTags = true,
  showCheckbox = false,

  // Infinite scroll options
  threshold = 0.1,
  rootMargin = "100px",
  className = "",
}) => {
  // Infinite scroll hook
  const { sentinelRef, hasReachedEnd } = useInfiniteScroll({
    threshold,
    rootMargin,
    enabled: !isLoading,
    hasNextPage,
    isFetchingNextPage: isLoadingMore,
    onLoadMore,
  });

  // Memoize newsletter row actions to prevent unnecessary re-renders
  const createNewsletterActions = useMemo(() => {
    return (newsletter: NewsletterWithRelations) => ({
      onToggleSelect: onToggleSelect
        ? () => onToggleSelect(newsletter.id)
        : undefined,
      onToggleLike: onToggleLike
        ? () => onToggleLike(newsletter)
        : async () => {},
      onToggleArchive: onToggleArchive
        ? async () => onToggleArchive(newsletter.id)
        : async () => {},
      onToggleRead: onToggleRead
        ? async () => onToggleRead(newsletter.id)
        : async () => {},
      onTrash: onTrash ? () => onTrash(newsletter.id) : () => {},
      onToggleQueue: onToggleQueue
        ? async () => onToggleQueue(newsletter.id)
        : async () => {},
      onUpdateTags: onUpdateTags
        ? (newsletterId: string, tagIds: string[]) =>
            onUpdateTags(newsletterId, tagIds)
        : () => {},
      onToggleTagVisibility: onToggleTagVisibility || (() => {}),
      onTagClick: onTagClick
        ? (tag: Tag, _e: React.MouseEvent) => onTagClick(tag)
        : () => {},
      onRemoveFromQueue: onRemoveFromQueue,
      onRowClick: onRowClick ? () => onRowClick(newsletter) : undefined,
      onMouseEnter: onMouseEnter ? () => onMouseEnter(newsletter) : undefined,
    });
  }, [
    onToggleSelect,
    onToggleLike,
    onToggleArchive,
    onToggleRead,
    onTrash,
    onToggleQueue,
    onUpdateTags,
    onToggleTagVisibility,
    onTagClick,
    onRemoveFromQueue,
    onRowClick,
    onMouseEnter,
  ]);

  // Initial loading state
  if (isLoading && newsletters.length === 0) {
    return (
      <div
        className={`flex flex-col items-center justify-center py-16 ${className}`}
      >
        <div className="animate-spin rounded-full h-12 w-12 border-t-2 border-b-2 border-blue-500 mb-4"></div>
        <p className="text-base text-neutral-400">Loading newsletters...</p>
      </div>
    );
  }

  // Error state for initial load
  if (error && newsletters.length === 0) {
    return (
      <div
        className={`flex flex-col items-center justify-center py-16 text-center ${className}`}
      >
        <div className="text-red-500 mb-4">
          <svg
            className="w-16 h-16 mx-auto mb-4"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
            />
          </svg>
          <p className="text-lg font-medium text-gray-900 mb-2">
            Failed to load newsletters
          </p>
          <p className="text-sm text-gray-500 mb-4">
            {error.message ||
              "Something went wrong while loading your newsletters"}
          </p>
        </div>
        {onRetry && (
          <button
            onClick={onRetry}
            className="px-6 py-2 bg-blue-600 hover:bg-blue-700 text-white rounded-lg transition-colors"
          >
            Try Again
          </button>
        )}
      </div>
    );
  }

  // Empty state
  if (!isLoading && newsletters.length === 0) {
    return (
      <div
        className={`flex flex-col items-center justify-center py-16 text-center ${className}`}
      >
        <div className="text-gray-400 mb-4">
          <svg
            className="w-16 h-16 mx-auto"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              d="M20 13V6a2 2 0 00-2-2H6a2 2 0 00-2 2v7m16 0v5a2 2 0 01-2 2H6a2 2 0 01-2 2m16-7H4m16 0l-2-2m-12 2l2-2"
            />
          </svg>
        </div>
        <h3 className="text-lg font-medium text-gray-900 mb-2">
          No newsletters found
        </h3>
        <p className="text-sm text-gray-500">
          No newsletters match your current filters. Try adjusting your search
          criteria.
        </p>
      </div>
    );
  }

  return (
    <div className={`${className}`}>
      {/* Newsletter List */}
      <div className="space-y-0 divide-y divide-gray-100">
        {newsletters.map((newsletter) => {
          const isInQueue = readingQueue.some(
            (item) => item.newsletter_id === newsletter.id,
          );

          const actions = createNewsletterActions(newsletter);

          return (
            <NewsletterRow
              key={newsletter.id}
              newsletter={newsletter}
              isSelected={isSelecting && selectedIds.has(newsletter.id)}
              showCheckbox={showCheckbox || isSelecting}
              showTags={showTags}
              isInReadingQueue={isInQueue}
              readingQueue={readingQueue}
              visibleTags={visibleTags}
              isDeletingNewsletter={
                isDeletingNewsletter
                  ? isDeletingNewsletter(newsletter.id)
                  : false
              }
              loadingStates={loadingStates}
              errorTogglingLike={errorTogglingLike}
              isUpdatingTags={
                isUpdatingTags ? isUpdatingTags(newsletter.id) : false
              }
              tagUpdateError={tagUpdateError}
              onDismissTagError={onDismissTagError}
              {...actions}
            />
          );
        })}
      </div>

      {/* Loading Sentinel */}
      <div ref={sentinelRef}>
        <LoadingSentinel
          isLoading={isLoadingMore}
          hasReachedEnd={hasReachedEnd}
          totalCount={totalCount}
          loadedCount={newsletters.length}
          error={error}
          onRetry={onRetry}
          className="mt-4"
        />
      </div>
    </div>
  );
};

export default InfiniteNewsletterList;
