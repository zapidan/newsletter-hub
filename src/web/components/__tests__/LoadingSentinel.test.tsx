import { render, screen } from '@testing-library/react';
import { vi } from 'vitest';
import { LoadingSentinel } from '../InfiniteScroll/LoadingSentinel';

describe('LoadingSentinel', () => {
  it('renders loading spinner when loading is true', () => {
    const { container } = render(
      <LoadingSentinel
        isLoading={true}
        hasReachedEnd={false}
        onRetry={() => { }}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Loading more newsletters...')).toBeInTheDocument();
    const spinner = container.querySelector('.animate-spin');
    expect(spinner).toBeInTheDocument();
    expect(spinner).toHaveClass('w-6', 'h-6', 'text-blue-500');
  });

  it('renders loading more indicator with counts when loading with counts', () => {
    const { container } = render(
      <LoadingSentinel
        isLoading={true}
        hasReachedEnd={false}
        onRetry={() => { }}
        loadedCount={10}
        totalCount={100}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Loading more newsletters...')).toBeInTheDocument();
    expect(screen.getByText('Loaded 10 of 100')).toBeInTheDocument();
  });

  it('renders error state when error is present', () => {
    // Use the same error message as in the snapshot
    const error = new Error('Failed to load');
    const handleRetry = vi.fn();

    const { container } = render(
      <LoadingSentinel
        isLoading={false}
        error={error}
        onRetry={handleRetry}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Failed to load more newsletters')).toBeInTheDocument();
    expect(screen.getByText('Failed to load')).toBeInTheDocument();

    const retryButton = screen.getByRole('button', { name: /try again/i });
    retryButton.click();
    expect(handleRetry).toHaveBeenCalledTimes(1);
  });

  it('renders end of content with counts when hasReachedEnd is true and counts are provided', () => {
    const { container } = render(
      <LoadingSentinel
        isLoading={false}
        hasReachedEnd={true}
        onRetry={() => { }}
        loadedCount={50}
        totalCount={50}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('All 50 newsletters loaded')).toBeInTheDocument();
    expect(screen.getByText('Showing 50 of 50 newsletters')).toBeInTheDocument();
  });

  it('renders basic end of content when hasReachedEnd is true without counts', () => {
    const { container } = render(
      <LoadingSentinel
        isLoading={false}
        hasReachedEnd={true}
        onRetry={() => { }}
      />
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText('No more newsletters to load')).toBeInTheDocument();
  });

  it('renders a hidden div when not loading and no error', () => {
    const { container } = render(
      <LoadingSentinel
        isLoading={false}
        hasReachedEnd={false}
        onRetry={() => { }}
      />
    );

    expect(container).toMatchSnapshot();
    const hiddenDiv = container.firstChild as HTMLElement;
    expect(hiddenDiv).toHaveClass('h-4', 'w-full');
    expect(hiddenDiv).toHaveAttribute('aria-hidden', 'true');
  });
});

beforeEach(() => {
  vi.clearAllMocks();
});
