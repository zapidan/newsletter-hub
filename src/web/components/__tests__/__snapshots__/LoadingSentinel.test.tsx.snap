// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoadingSentinel > renders a hidden div when not loading and no error 1`] = `
<div>
  <div
    aria-hidden="true"
    class="h-4 w-full "
  />
</div>
`;

exports[`LoadingSentinel > renders basic end of content when hasReachedEnd is true without counts 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center py-8 text-center "
  >
    <div
      class="text-gray-400 mb-2"
    >
      <svg
        class="w-8 h-8 mx-auto"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="text-sm text-gray-500 font-medium"
    >
      No more newsletters to load
    </p>
  </div>
</div>
`;

exports[`LoadingSentinel > renders end of content with counts when hasReachedEnd is true and counts are provided 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center py-8 text-center "
  >
    <div
      class="text-gray-400 mb-2"
    >
      <svg
        class="w-8 h-8 mx-auto"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
    <p
      class="text-sm text-gray-500 font-medium"
    >
      All 50 newsletters loaded
    </p>
    <p
      class="text-xs text-gray-400 mt-1"
    >
      Showing 
      50
       of 
      50
       newsletters
    </p>
  </div>
</div>
`;

exports[`LoadingSentinel > renders error state when error is present 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center py-8 text-center "
  >
    <div
      class="text-red-500 mb-4"
    >
      <svg
        class="w-12 h-12 mx-auto mb-2"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 9v2m0 4h.01m-6.938 4h13.856c1.54 0 2.502-1.667 1.732-2.5L13.732 4c-.77-.833-1.964-.833-2.732 0L3.732 16.5c-.77.833.192 2.5 1.732 2.5z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      <p
        class="text-sm font-medium"
      >
        Failed to load more newsletters
      </p>
      <p
        class="text-xs text-gray-500 mt-1"
      >
        Failed to load
      </p>
    </div>
    <button
      class="px-4 py-2 text-sm bg-red-50 hover:bg-red-100 text-red-700 rounded-lg transition-colors"
    >
      Try Again
    </button>
  </div>
</div>
`;

exports[`LoadingSentinel > renders loading more indicator with counts when loading with counts 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center py-8 "
  >
    <svg
      class="lucide lucide-loader2 w-6 h-6 animate-spin text-blue-500 mb-3"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 12a9 9 0 1 1-6.219-8.56"
      />
    </svg>
    <p
      class="text-sm text-gray-500 font-medium"
    >
      Loading more newsletters...
    </p>
    <p
      class="text-xs text-gray-400 mt-1"
    >
      Loaded 
      10
       of 
      100
    </p>
  </div>
</div>
`;

exports[`LoadingSentinel > renders loading spinner when loading is true 1`] = `
<div>
  <div
    class="flex flex-col items-center justify-center py-8 "
  >
    <svg
      class="lucide lucide-loader2 w-6 h-6 animate-spin text-blue-500 mb-3"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 12a9 9 0 1 1-6.219-8.56"
      />
    </svg>
    <p
      class="text-sm text-gray-500 font-medium"
    >
      Loading more newsletters...
    </p>
  </div>
</div>
`;
